proposals/csharp-10.0/parameterless-struct-constructors.md

```diff
@@ -0,0 +1,102 @@
+# Parameterless struct constructors
+
+## Summary
+
+Support explicit parameterless constructors for value types.
+
+## Proposal
+
+If the parameterless constructor is less accessible than the value type, a warning is reported and the constructor is ignored when inaccessible.
+```csharp
+struct NoConstructor
+{
+    // ok
+}
+struct PublicConstructor
+{
+    internal PublicConstructor() { } // ok: always accessible
+}
+struct PrivateConstructor
+{
+    private PrivateConstructor() { } // warning: parameterless .ctor will be ignored if inaccessible
+}
+```
+
+`default(S)` ignores any parameterless constructor and generates a zeroed `S`.
+```csharp
+NoConstructor x = default;      // ok
+PublicConstructor y = default;  // ok
+PrivateConstructor z = default; // ok
+```
+
+For a type parameter `T` with a `new()` constraint, `new T()` is emitted as a call to `System.Activator.CreateInstance<T>()` which ignores any explicit parameterless constructor.
+A warning is reported if there is an accessible parameterless constructor.
+
+_Should `Activator.CreateInstance()` call the parameterless constructor if available and accessible?_
+```csharp
+static T Create<T>() where T : new() => new T();
+
+x = Create<NoConstructor>();      // ok
+y = Create<PublicConstructor>();  // warning: constructor ignored
+z = Create<PrivateConstructor>(); // ok
+```
+
+A local or field of value type `S` that is not explicitly initialized is zeroed.
+The compiler already reports `error: use of unassigned local` for uninitialized locals of a value type that is not empty, and there is no additional handling beyond that for value types with explicit parameterless constructors.
+```csharp
+NoConstructor x;
+PublicConstructor y;
+PrivateConstructor z;
+x.ToString(); // error: use of unassigned local (unless type is empty)
+y.ToString(); // error: use of unassigned local (unless type is empty)
+z.ToString(); // error: use of unassigned local (unless type is empty)
+```
+
+Array allocation ignores any parameterless constructor and generates zeroed elements.
+A warning is reported if there is an accessible parameterless constructor.
+```csharp
+var a = new NoConstructor[1];      // ok
+var b = new PublicConstructor[1];  // warning: constructor ignored
```

Comment:
Is a warning for this specific case needed, given the expectation that array creation never runs constructors for elements? Or conversely, should there be a warning for:

```cs
var b = new ClassWithPublicConstructor[1];
```

Reply:
Also seems a tad extraneous given that C# has always supported the consumption of structs with parameterless constructors and the behavior here will not change, but it's understandable to be explicit given how exceptionally unlikely a C# dev will have run into such a struct before.

Reply:
Makes sense. Because the idea that newing up an array would execute constructors runs so counter to how I think about types both with and without default constructors, it would be nice if this was a separate warning code so that I could suppress just the new array scenario without suppressing the other checks.

Reply:
Where I'm coming from is wanting to treat structs more like classes in this regard, since one of the main reasons I don't go with a struct these days is if I want a default constructor. So when I picture what it will be like when structs with parameterless constructors are common, I want to reason about constructors the same way.
But the difference between structs and classes will never be completely dismissable. If I think about the `new PublicConstructor[1]` case in terms of filling the array with Formatter.GetUninitializedObject, suddenly warnings feel really good.

So I get that, but where I still land is that there's really no other way to new up an array of structs that have parameterless constructors, even if you do immediately fill the array with valid instances right after you create the array. I'm picturing that there will be a lot of rote suppressions in my future.

Reply:
This warning doesn't feel correct. We don't warn for classes, the same should apply to structs.

Reply:
> there's really no other way to new up an array of structs that have parameterless constructors

Since there's no way to avoid the warning, I've removed the warning comment and added an open issue.

---
In reply to: [581504602](https://github.com/dotnet/csharplang/pull/4459#discussion_r581504602) [](ancestors = 581504602)